src/compiler/checker.ts

```diff
@@ -27160,6 +27160,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             case SyntaxKind.ElementAccessExpression:
                 // The resolvedSymbol property is initialized by checkPropertyAccess or checkElementAccess before we get here.
                 return isConstantReference((node as AccessExpression).expression) && isReadonlySymbol(getNodeLinks(node).resolvedSymbol || unknownSymbol);
+            case SyntaxKind.ObjectBindingPattern:
+            case SyntaxKind.ArrayBindingPattern:
+                return isVariableDeclaration(node.parent) && isVarConstLike(node.parent);

```

Comment:
shouldn't this lead to calling `isConstantReference` recursively (perhaps accompanied by some other checks but still)? What about nested binding patterns (`const { a: { b: b2 } } = ...`)?

Reply:
Thank you for bringing this to my attention! I was surprised to discover that TypeScript currently lacks support for nested binding patterns during cross symbol analysis. I don't know if it is intended though.

[playground](https://www.typescriptlang.org/play?useDefineForClassFields=true&target=9&module=1&noImplicitUseStrict=true&ts=5.3.0-dev.20231029#code/C4TwDgpgBAchDOwIBMoF4oG8BQU9VEgC4oByRAJwEsA7Ac1IG5d8KEwSd9upkBDYHxKVadFngC+2CVAA+WcQXAQSpGgFcAtgCMIFJorbwOCnvn6CSGnXubcpU7F3wAbCMChGwAQhJxEKHb4AMYA9jSIWJ7snLwCfFASADRKkIno0cZBeFQAZlAAFITQaKVkIvSkAJSmZnGCUPACVPC5VAiNwNT0jHgA9H2NABah6i6oNKEeehShFIpS3HmFXgB0xehl5F2i1bVma2sWCU3ALW0dFXTZidLYQA)

```ts
type Nested = {
    type: 'string';
    resp: {
        data: string
    }
} | {
    type: 'number';
    resp: {
        data: number;
    }
}

{
    let resp!: Nested;
    const { resp: { data }, type } = resp;
    if (type === 'string') {
        data satisfies string;  // should not error
    }
    if (resp.type === 'string') {
        resp.resp.data satisfies string;
    }
}
```

Reply:
This sounds like https://github.com/microsoft/TypeScript/pull/38839#issuecomment-1160929515

I was more thinking about cases like this:
```ts
type Nested =
  | {
      resp: {
        type: "string";
        data: string;
      };
    }
  | {
      resp: {
        type: "number";
        data: number;
      };
    };

declare const nested: Nested;

const {
  resp: { type, data },
} = nested;
if (type === "string") {
  data satisfies string; // should be OK
}
```

This keeps the dependency between the discriminant and another property "at the same level" but at the same time both are coming from a nested binding pattern.

Reply:
I think that I might have accidentally fixed this in [this PR](https://github.com/microsoft/TypeScript/pull/56306). When I sat down to it I totally forgot about the conversation here 😅